### PR TITLE
ocrvs-2458 Reinstate archived application

### PR DESCRIPTION
**Pull request by sadmananik**

### CODE REVIEW

**Comment by sadmananik** [comment]
This PR has a deoendency PR on Farajaland

**Comment by sadmananik** [comment]
ocrvs-2458

**Comment by sadmananik** [comment]
This issue has a dependency PR on [[fara](https://github.com/opencrvs/opencrvs-farajaland/pull/182)](https://github.com/opencrvs/opencrvs-farajaland/pull/182)

**Review by yeasinhossain-dsi — APPROVED** [review]

File: packages/client/src/SubmissionController.test.ts [hunk 1]
```diff
@@ -576,14 +585,114 @@ describe('Submission Controller', () => {
     subCon.client.setRequestHandler(ARCHIVE_BIRTH_DECLARATION, mutationHandler)
 
     await subCon.sync()
-
+    await new Promise((resolve) => {
+      setTimeout(resolve, 500)
+    })
```
Comment (Zangetsu101) [inline packages/client/src/SubmissionController.test.ts:590]: Don't think this is needed here

File: packages/client/src/SubmissionController.ts [hunk 2]
```diff
@@ -78,6 +82,9 @@ const ACTION_LIST: IActionList = {
   [Action.APPROVE_APPLICATION]: Action.APPROVE_APPLICATION,
   [Action.REGISTER_APPLICATION]: Action.REGISTER_APPLICATION,
   [Action.REJECT_APPLICATION]: Action.REJECT_APPLICATION,
+  [Action.ARCHIVE_APPLICATION]: Action.ARCHIVE_APPLICATION,
+  [Action.REINSTATE_APPLICATION]: Action.REINSTATE_APPLICATION,
+  [Action.COLLECT_CERTIFICATE]: Action.COLLECT_CERTIFICATE,
   [Action.ARCHIVE_DECLARATION]: Action.ARCHIVE_DECLARATION,
   [Action.COLLECT_CERTIFICATE]: Action.COLLECT_CERTIFICATE,
```
Comment (Zangetsu101) [inline packages/client/src/SubmissionController.ts:89]: `COLLECT_CERTIFICATE` is mapped twice here

File: packages/client/src/applications/index.ts [hunk 3]
```diff
@@ -319,6 +323,12 @@ interface IDeleteApplicationAction {
     application: IApplication | IPrintableApplication
   } & OnSuccessDeleteApplicationOptions
 }
+interface IREINSTATE_APPLICATIONAction {
```
Comment (Zangetsu101) [inline packages/client/src/applications/index.ts:326]: `IReinstateApplicationAction` seems more appropriate

File: packages/client/src/applications/index.ts [hunk 4]
```diff
@@ -414,6 +424,7 @@ export type Action =
   | IWriteApplicationAction
   | NavigationAction
   | IDeleteApplicationAction
+  | IREINSTATE_APPLICATIONAction
```
Comment (Zangetsu101) [inline packages/client/src/applications/index.ts:427]: Same as above

File: packages/client/src/forms/index.ts [hunk 5]
```diff
@@ -77,9 +77,11 @@ export enum Action {
   REGISTER_APPLICATION = 'register',
   COLLECT_CERTIFICATE = 'collect certificate',
   REJECT_APPLICATION = 'reject',
-  ARCHIVE_DECLARATION = 'archive',
+  ARCHIVE_APPLICATION = 'archive',
```
Comment (Zangetsu101) [inline packages/client/src/forms/index.ts:80]: This should have been deleted as we already have `ARCHIVE_DECLARATION` here

File: packages/client/src/SubmissionController.test.ts [hunk 6]
```diff
@@ -20,14 +20,16 @@ import {
   REGISTER_BIRTH_APPLICATION,
   REJECT_BIRTH_APPLICATION,
   COLLECT_BIRTH_CERTIFICATE,
-  ARCHIVE_BIRTH_DECLARATION
+  ARCHIVE_BIRTH_DECLARATION,
+  REINSTATE_BIRTH_APPLICATION
```
Comment (Zangetsu101) [inline packages/client/src/SubmissionController.test.ts:24]: We need to refactor `REINSTATE_BIRTH_APPLICATION` to `REINSTATE_BIRTH_DECLARATION`

File: packages/client/src/SubmissionController.test.ts [hunk 7]
```diff
@@ -20,14 +20,16 @@ import {
...
 } from '@client/views/DataProvider/birth/mutations'
 import {
   ARCHIVE_DEATH_DECLARATION,
   APPROVE_DEATH_APPLICATION,
   COLLECT_DEATH_CERTIFICATE,
   REGISTER_DEATH_APPLICATION,
-  REJECT_DEATH_APPLICATION
+  REJECT_DEATH_APPLICATION,
+  REINSTATE_DEATH_APPLICATION
```
Comment (Zangetsu101) [inline packages/client/src/SubmissionController.test.ts:32]: We need to refactor `REINSTATE_DEATH_APPLICATION` to `REINSTATE_DEATH_DECLARATION`

File: packages/client/src/views/DataProvider/death/mutations.ts [hunk 8]
```diff
@@ -76,6 +76,19 @@ export const REJECT_DEATH_APPLICATION = gql`
...
+      taskEntryResourceID
+      registrationStatus
+    }
+  }
+`
+export const ARCHIVE_DEATH_APPLICATION = gql`
+  mutation submitMutation($id: String!) {
+    markEventAsArchived(id: $id)
+  }
+`
```
Comment (Zangetsu101) [inline packages/client/src/views/DataProvider/death/mutations.ts:91]: This needs to be removed

File: packages/client/src/views/Home/RecordAudit.tsx [hunk 9]
```diff
@@ -246,6 +254,18 @@ const STATUSTOCOLOR: { [key: string]: string } = {
...
+  type: 'Event',
+  trackingId: 'Tracking ID',
+  dateOfBirth: 'Date of birth',
+  dateOfDeath: 'Date of death',
+  placeOfBirth: 'Place of birth',
+  placeOfDeath: 'Place of death',
+  informant: 'Informant',
+  brn: 'BRN',
+  drn: 'DRN'
+}
```
Comment (Zangetsu101) [inline packages/client/src/views/Home/RecordAudit.tsx:268]: This is redundant

File: packages/client/src/views/OfficeHome/NotificationToast.tsx [hunk 10]
```diff
@@ -53,11 +53,14 @@ class NotificationToast extends React.Component<IFullProps> {
...
         item.submissionStatus === SUBMISSION_STATUS.READY_TO_SUBMIT ||
         item.submissionStatus === SUBMISSION_STATUS.READY_TO_APPROVE ||
         item.submissionStatus === SUBMISSION_STATUS.READY_TO_REGISTER ||
         item.submissionStatus === SUBMISSION_STATUS.READY_TO_REJECT ||
         item.submissionStatus === SUBMISSION_STATUS.READY_TO_CERTIFY ||
+        item.submissionStatus === SUBMISSION_STATUS.ARCHIVING ||
+        item.submissionStatus === SUBMISSION_STATUS.REINSTATING ||
         item.submissionStatus ===
           SUBMISSION_STATUS.READY_TO_REQUEST_CORRECTION ||
         item.submissionStatus === SUBMISSION_STATUS.ARCHIVING ||
```
Comment (Zangetsu101) [inline packages/client/src/views/OfficeHome/NotificationToast.tsx:66]: `SUBMISSION_STATUS.ARCHIVING` is used twice here

File: packages/gateway/src/features/registration/root-resolvers.test.ts [hunk 11]
```diff
@@ -1023,6 +1023,273 @@ describe('Registration root resolvers', () => {
...
+                  },
+                  meta: {
+                    lastUpdated: '2018-11-29T10:40:08.913+00:00',
+                    versionId: 'aa8c1c4a-4680-497f-81f7-fde357fdb77d'
+                  },
+                  id: 'ba0412c6-5125-4447-bd32-fb5cf336ddbc'
+                }
+              }
+            ]
+          })
```
Comment (Zangetsu101) [inline packages/gateway/src/features/registration/root-resolvers.test.ts:1101]: `mockTestBundle` might be reused here

File: packages/gateway/src/features/registration/root-resolvers.test.ts [hunk 12]
```diff
@@ -1023,6 +1023,273 @@ describe('Registration root resolvers', () => {
...
+      expect(result).toBe('ba0412c6-5125-4447-bd32-fb5cf336ddbc')
+    })
+
+    it('throws error if user does not have register or validate scope', async () => {
+      const id = 'df3fb104-4c2c-486f-97b3-edbeabcd4422'
+      await expect(
+        resolvers.Mutation.markEventAsArchived({}, { id }, authHeaderNotRegCert)
+      ).rejects.toThrowError('User does not have a register or validate scope')
+    })
+  })
```
Comment (Zangetsu101) [inline packages/gateway/src/features/registration/root-resolvers.test.ts:1291]: This is already defined above once

File: packages/gateway/src/features/registration/root-resolvers.ts [hunk 13]
```diff
@@ -362,6 +363,76 @@ export const resolvers: GQLResolver = {
...
+          authHeader
+        )
+
+        const taskEntryData = taskBundle.entry[0]
+        if (!taskEntryData) {
+          throw new Error('Task does not exist')
+        }
+
+        const taskEntryResourceID =
+          taskBundle.entry && taskBundle.entry[0].resource.id
```
Comment (Zangetsu101) [inline packages/gateway/src/features/registration/root-resolvers.ts:382]: `taskEntryData.resource.id`

File: packages/client/src/SubmissionController.test.ts (hunk 1 above)
Comment (sadmananik) [reply packages/client/src/SubmissionController.test.ts:590]: Removed

File: packages/client/src/SubmissionController.ts (hunk 2 above)
Comment (sadmananik) [reply packages/client/src/SubmissionController.ts:89]: Removed

File: packages/client/src/applications/index.ts (hunk 3 above)
Comment (sadmananik) [reply packages/client/src/applications/index.ts:326]: Changed

File: packages/client/src/applications/index.ts (hunk 4 above)
Comment (sadmananik) [reply packages/client/src/applications/index.ts:427]: Changed

File: packages/client/src/forms/index.ts (hunk 5 above)
Comment (sadmananik) [reply packages/client/src/forms/index.ts:80]: Removed

File: packages/client/src/SubmissionController.test.ts (hunk 6 above)
Comment (sadmananik) [reply packages/client/src/SubmissionController.test.ts:24]: Changed

File: packages/client/src/SubmissionController.test.ts (hunk 7 above)
Comment (sadmananik) [reply packages/client/src/SubmissionController.test.ts:32]: Changed

File: packages/client/src/views/DataProvider/death/mutations.ts (hunk 8 above)
Comment (sadmananik) [reply packages/client/src/views/DataProvider/death/mutations.ts:91]: Removed

File: packages/client/src/views/Home/RecordAudit.tsx (hunk 9 above)
Comment (sadmananik) [reply packages/client/src/views/Home/RecordAudit.tsx:268]: Removed

File: packages/gateway/src/features/registration/root-resolvers.test.ts (hunk 12 above)
Comment (sadmananik) [reply packages/gateway/src/features/registration/root-resolvers.test.ts:1291]: Removed

File: packages/gateway/src/features/registration/root-resolvers.ts (hunk 13 above)
Comment (sadmananik) [reply packages/gateway/src/features/registration/root-resolvers.ts:382]: Changed

File: packages/gateway/src/features/registration/root-resolvers.test.ts (hunk 11 above)
Comment (sadmananik) [reply packages/gateway/src/features/registration/root-resolvers.test.ts:1101]: seems logical. Too much work.

File: packages/search/src/elasticsearch/utils.ts [hunk 14]
```diff
@@ -424,7 +431,9 @@ export function isValidOperationHistory(body: IBirthCompositionBody) {
       WAITING_VALIDATION_STATUS
     ],
     [CERTIFIED_STATUS]: [REGISTERED_STATUS, CERTIFIED_STATUS],
-    [REQUESTED_CORRECTION_STATUS]: [REGISTERED_STATUS, CERTIFIED_STATUS]
+    [REQUESTED_CORRECTION_STATUS]: [REGISTERED_STATUS, CERTIFIED_STATUS],
+    [REINSTATED_STATUS]: [ARCHIVED_STATUS],
+    [CERTIFIED_STATUS]: [REGISTERED_STATUS, CERTIFIED_STATUS]
```
[the remark this answers is not in the document]
Comment (sadmananik) [reply packages/search/src/elasticsearch/utils.ts:436]: Without these it gives error while searching for the first time login.

**Pull request merged**